File: src/agent/simulator.py
"""Simulador de compras (carro / imóvel).

Compara duas estratégias para uma compra:
1. **Financiar** — entrada agora + parcelas. Suporta tabela Price (parcela fixa)
   e SAC (amortização constante, parcela decrescente — comum em imóvel).
2. **Juntar à vista** — guardar a sobra mensal até ter o valor total (com ou sem
   rendimento), e comprar sem juros.

Funções puras, sem dependência de banco — fáceis de testar.
"""

from __future__ import annotations

from typing import Any


def _taxa_mensal(juros_aa: float) -> float:
    """Converte taxa anual efetiva (%) em taxa mensal efetiva (fração)."""
    return (1 + juros_aa / 100) ** (1 / 12) - 1


def simular_price(valor_financiado: float, prazo_meses: int, juros_aa: float) -> dict[str, Any]:
    i = _taxa_mensal(juros_aa)
    n = max(1, int(prazo_meses))
    if i <= 0:
        parcela = valor_financiado / n
    else:
        parcela = valor_financiado * i / (1 - (1 + i) ** (-n))
    total = parcela * n
    return {
        "sistema": "price",
        "parcela": round(parcela, 2),
        "total_parcelas": round(total, 2),
        "total_juros": round(total - valor_financiado, 2),
    }


def simular_sac(valor_financiado: float, prazo_meses: int, juros_aa: float) -> dict[str, Any]:
    i = _taxa_mensal(juros_aa)
    n = max(1, int(prazo_meses))
    amortizacao = valor_financiado / n
    primeira = amortizacao + valor_financiado * i
    ultima = amortizacao + amortizacao * i
    total_juros = i * valor_financiado * (n + 1) / 2  # soma PA dos juros
    return {
        "sistema": "sac",
        "primeira_parcela": round(primeira, 2),
        "ultima_parcela": round(ultima, 2),
        "total_parcelas": round(valor_financiado + total_juros, 2),
        "total_juros": round(total_juros, 2),
    }


def simular_consorcio(preco: float, prazo_meses: int, taxa_adm_pct: float = 18.0) -> dict[str, Any]:
    """Consórcio: sem juros, mas com taxa de administração sobre a carta (simplificado).

    Não há uso imediato do bem até a contemplação — modelado só no custo.
    """
    n = max(1, int(prazo_meses))
    total = preco * (1 + taxa_adm_pct / 100)
    return {
        "sistema": "consorcio",
        "taxa_adm_pct": taxa_adm_pct,
        "parcela": round(total / n, 2),
        "total_parcelas": round(total, 2),
        "custo_taxa_adm": round(total - preco, 2),
    }


def _payoff(saldo: float, i: float, parcela: float, teto: int = 1200) -> dict[str, Any] | None:
    """Quita um saldo pagando `parcela` fixa/mês a juros `i`. None se nunca quita."""
    juros_pagos = 0.0
    s = saldo
    for m in range(1, teto + 1):
        juros = s * i
        amort = parcela - juros
        if amort <= 0:
            return None  # parcela não cobre os juros
        juros_pagos += juros
        s -= amort
        if s <= 0:
            return {"meses": m, "juros_pagos": round(juros_pagos, 2)}
    return None


def simular_amortizacao(
    saldo: float, juros_aa: float, parcela: float, aporte_extra: float = 0.0
) -> dict[str, Any]:
    """Compara quitar um financiamento com a parcela atual vs. com aporte extra/mês."""
    i = _taxa_mensal(juros_aa)
    base = _payoff(saldo, i, parcela)
    com = _payoff(saldo, i, parcela + aporte_extra) if aporte_extra > 0 else base
    economia = None
    if base and com:
        economia = round(base["juros_pagos"] - com["juros_pagos"], 2)
    return {
        "saldo": round(saldo, 2),
        "parcela_atual": round(parcela, 2),
        "aporte_extra": round(aporte_extra, 2),
        "sem_extra": base,
        "com_extra": com,
        "meses_economizados": (base["meses"] - com["meses"]) if (base and com) else None,
        "juros_economizados": economia,
    }


def meses_para_juntar(
    alvo: float, aporte_mensal: float, rendimento_aa: float = 0.0, teto: int = 1200
) -> int | None:
    """Quantos meses para acumular `alvo` guardando `aporte_mensal` (com rendimento)."""
    if aporte_mensal <= 0 or alvo <= 0:
        return None
    r = _taxa_mensal(rendimento_aa)
    saldo = 0.0
    for m in range(1, teto + 1):
        saldo = saldo * (1 + r) + aporte_mensal
        if saldo >= alvo:
            return m
    return None  # não atinge dentro do teto


def simular_compra(
    *,
    preco: float,
    entrada: float = 0.0,
    prazo_meses: int = 48,
    juros_aa: float = 18.0,
    sobra_mensal: float = 0.0,
    rendimento_aa: float = 0.0,
    taxa_adm_consorcio: float = 18.0,
) -> dict[str, Any]:
    """Compara financiar (Price + SAC), consórcio e juntar à vista."""
    preco = float(preco)
    entrada = min(float(entrada), preco)
    valor_financiado = max(0.0, preco - entrada)

    price = simular_price(valor_financiado, prazo_meses, juros_aa)
    sac = simular_sac(valor_financiado, prazo_meses, juros_aa)
    consorcio = simular_consorcio(preco, prazo_meses, taxa_adm_consorcio)
    custo_financiar_price = round(entrada + price["total_parcelas"], 2)
    custo_financiar_sac = round(entrada + sac["total_parcelas"], 2)

    meses_juntar = meses_para_juntar(preco, sobra_mensal, rendimento_aa)
    juntar = {
        "aporte_mensal": round(sobra_mensal, 2),
        "rendimento_aa": rendimento_aa,
        "meses_para_juntar": meses_juntar,
        "anos_para_juntar": round(meses_juntar / 12, 1) if meses_juntar else None,
        "custo_total": round(preco, 2),  # à vista: sem juros
    }

    # Economia de juros ao juntar à vista vs financiar (Price).
    economia_vs_price = round(custo_financiar_price - preco, 2)

    return {
        "entrada": round(entrada, 2),
        "valor_financiado": round(valor_financiado, 2),
        "financiar": {
            "price": price,
            "sac": sac,
            "custo_total_price": custo_financiar_price,
            "custo_total_sac": custo_financiar_sac,
        },
        "consorcio": consorcio,
        "juntar_a_vista": juntar,
        "economia_juntando_vs_price": economia_vs_price,
    }

File: web/lib/simulator.ts
import { formatBRL } from "@/lib/format";
import type {
  AmortizationRequest,
  AmortizationResponse,
  MaintenanceResponse,
  ScenarioSimulationRequest,
  ScenarioSimulationResponse,
} from "@/lib/types";

export type ValidationResult = { valid: true } | { valid: false; message: string };
export type ResultTone = "neutral" | "positive" | "negative";

export type ScenarioPresetKey = "carro" | "imovel";

export type ScenarioSummaryItem = {
  label: string;
  value: string;
  tone: ResultTone;
};

export type ScenarioSummaryCard = {
  key: "financiar" | "consorcio" | "juntar";
  title: string;
  subtitle: string;
  items: ScenarioSummaryItem[];
};

export type AmortizationRow = {
  key: "sem_extra" | "com_extra" | "economia";
  label: string;
  months: string;
  interest: string;
  tone: ResultTone;
};

export const DEFAULT_SCENARIO_INPUT: ScenarioSimulationRequest = {
  preco: 50000,
  entrada: 10000,
  prazo_meses: 48,
  juros_aa: 24,
  sobra_mensal: 0,
  rendimento_aa: 10,
  taxa_adm_consorcio: 18,
};

export const SCENARIO_PRESETS: Record<ScenarioPresetKey, ScenarioSimulationRequest> = {
  carro: {
    ...DEFAULT_SCENARIO_INPUT,
    preco: 80000,
    entrada: 20000,
    prazo_meses: 48,
    juros_aa: 26,
    rendimento_aa: 11,
  },
  imovel: {
    ...DEFAULT_SCENARIO_INPUT,
    preco: 300000,
    entrada: 60000,
    prazo_meses: 360,
    juros_aa: 9,
    rendimento_aa: 11,
  },
};

export const DEFAULT_AMORTIZATION_INPUT: AmortizationRequest = {
  saldo: 32200.6,
  juros_aa: 4.5,
  parcela: 395,
  aporte_extra: 500,
};

function finiteNumber(value: number): boolean {
  return Number.isFinite(value);
}

function monthlyRate(annualPercent: number): number {
  return (1 + annualPercent / 100) ** (1 / 12) - 1;
}

function monthsText(months: number | null | undefined): string {
  return typeof months === "number" ? `${months} meses` : "não quita";
}

function timeToSaveText(months: number | null, years: number | null): string {
  if (!months) {
    return "não atinge";
  }
  if (!years) {
    return `${months} meses`;
  }
  return `${months} meses (~${years.toLocaleString("pt-BR", { maximumFractionDigits: 1 })} anos)`;
}

export function validateScenarioInput(input: ScenarioSimulationRequest): ValidationResult {
  if (!finiteNumber(input.preco) || input.preco <= 0) {
    return { valid: false, message: "Informe um preço maior que zero." };
  }
  if (!finiteNumber(input.entrada) || input.entrada < 0) {
    return { valid: false, message: "Informe uma entrada válida." };
  }
  if (input.entrada > input.preco) {
    return { valid: false, message: "A entrada não pode ser maior que o preço." };
  }
  if (!finiteNumber(input.prazo_meses) || input.prazo_meses < 1) {
    return { valid: false, message: "Informe um prazo de pelo menos 1 mês." };
  }
  if (!finiteNumber(input.juros_aa) || input.juros_aa < 0) {
    return { valid: false, message: "Informe uma taxa de juros válida." };
  }
  if (!finiteNumber(input.sobra_mensal) || input.sobra_mensal < 0) {
    return { valid: false, message: "A sobra mensal não pode ser negativa." };
  }
  if (!finiteNumber(input.rendimento_aa) || input.rendimento_aa < 0) {
    return { valid: false, message: "O rendimento não pode ser negativo." };
  }
  if (!finiteNumber(input.taxa_adm_consorcio) || input.taxa_adm_consorcio < 0) {
    return { valid: false, message: "A taxa de administração não pode ser negativa." };
  }
  return { valid: true };
}

export function validateAmortizationInput(input: AmortizationRequest): ValidationResult {
  if (!finiteNumber(input.saldo) || input.saldo <= 0) {
    return { valid: false, message: "Informe um saldo devedor maior que zero." };
  }
  if (!finiteNumber(input.juros_aa) || input.juros_aa < 0) {
    return { valid: false, message: "Informe uma taxa de juros válida." };
  }
  if (!finiteNumber(input.parcela) || input.parcela <= 0) {
    return { valid: false, message: "Informe uma parcela maior que zero." };
  }
  if (!finiteNumber(input.aporte_extra) || input.aporte_extra < 0) {
    return { valid: false, message: "O aporte extra não pode ser negativo." };
  }

  const firstMonthInterest = input.saldo * monthlyRate(input.juros_aa);
  if (firstMonthInterest > 0 && input.parcela <= firstMonthInterest) {
    return { valid: false, message: "A parcela precisa cobrir os juros do primeiro mês." };
  }

  return { valid: true };
}

export function scenarioSummary(result: ScenarioSimulationResponse): ScenarioSummaryCard[] {
  return [
    {
      key: "financiar",
      title: "Financiar",
      subtitle: `Valor financiado ${formatBRL(result.valor_financiado)} com entrada de ${formatBRL(result.entrada)}.`,
      items: [
        {
          label: "Parcela Price",
          value: `${formatBRL(result.financiar.price.parcela)}/mês`,
          tone: "neutral",
        },
        {
          label: "Total pago Price",
          value: formatBRL(result.financiar.custo_total_price),
          tone: "negative",
        },
        {
          label: "Juros Price",
          value: formatBRL(result.financiar.price.total_juros),
          tone: "negative",
        },
        {
          label: "SAC 1ª → última",
          value: `${formatBRL(result.financiar.sac.primeira_parcela)} → ${formatBRL(result.financiar.sac.ultima_parcela)}`,
          tone: "neutral",
        },
        {
          label: "Total pago SAC",
          value: formatBRL(result.financiar.custo_total_sac),
          tone: "negative",
        },
      ],
    },
    {
      key: "consorcio",
      title: "Consórcio",
      subtitle: `Sem juros, com taxa administrativa de ${result.consorcio.taxa_adm_pct}%.`,
      items: [
        {
          label: "Parcela",
          value: `${formatBRL(result.consorcio.parcela)}/mês`,
          tone: "neutral",
        },
        {
          label: "Total pago",
          value: formatBRL(result.consorcio.total_parcelas),
          tone: "negative",
        },
        {
          label: "Custo da taxa",
          value: formatBRL(result.consorcio.custo_taxa_adm),
          tone: "negative",
        },
      ],
    },
    {
      key: "juntar",
      title: "Juntar à vista",
      subtitle: `Guardando ${formatBRL(result.juntar_a_vista.aporte_mensal)}/mês a ${result.juntar_a_vista.rendimento_aa}% a.a.`,
      items: [
        {
          label: "Tempo para juntar",
          value: timeToSaveText(
            result.juntar_a_vista.meses_para_juntar,
            result.juntar_a_vista.anos_para_juntar,
          ),
          tone: "neutral",
        },
        {
          label: "Custo à vista",
          value: formatBRL(result.juntar_a_vista.custo_total),
          tone: "positive",
        },
        {
          label: "Economia vs. Price",
          value: formatBRL(result.economia_juntando_vs_price),
          tone: "positive",
        },
      ],
    },
  ];
}

export function amortizationRows(result: AmortizationResponse): AmortizationRow[] {
  return [
    {
      key: "sem_extra",
      label: "Sem aporte extra",
      months: monthsText(result.sem_extra?.meses),
      interest: result.sem_extra ? formatBRL(result.sem_extra.juros_pagos) : "não quita",
      tone: "neutral",
    },
    {
      key: "com_extra",
      label: `Com aporte de ${formatBRL(result.aporte_extra)}/mês`,
      months: monthsText(result.com_extra?.meses),
      interest: result.com_extra ? formatBRL(result.com_extra.juros_pagos) : "não quita",
      tone: "positive",
    },
    {
      key: "economia",
      label: "Você economiza",
      months: monthsText(result.meses_economizados),
      interest:
        typeof result.juros_economizados === "number"
          ? formatBRL(result.juros_economizados)
          : "sem economia",
      tone: "positive",
    },
  ];
}

export function propertyToAmortizationInput(
  maintenance: MaintenanceResponse | null | undefined,
): AmortizationRequest | null {
  const property = maintenance?.family_profile.patrimonio?.imoveis?.[0];
  if (!property) {
    return null;
  }

  const saldo = Number(property.saldo_devedor ?? DEFAULT_AMORTIZATION_INPUT.saldo);
  const juros = Number(property.taxa_juros_anual ?? DEFAULT_AMORTIZATION_INPUT.juros_aa);
  const parcela = Number(
    property.custos?.financiamento ?? DEFAULT_AMORTIZATION_INPUT.parcela,
  );

  return {
    saldo: finiteNumber(saldo) && saldo > 0 ? saldo : DEFAULT_AMORTIZATION_INPUT.saldo,
    juros_aa: finiteNumber(juros) && juros >= 0 ? juros : DEFAULT_AMORTIZATION_INPUT.juros_aa,
    parcela: finiteNumber(parcela) && parcela > 0 ? parcela : DEFAULT_AMORTIZATION_INPUT.parcela,
    aporte_extra: DEFAULT_AMORTIZATION_INPUT.aporte_extra,
  };
}
